### PR TITLE
TST: migrate macOS CI to arm64, move x86 to cron

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

In the wake of #16319 and #16320, it seems we can no longer assume that both macOS' taget archs (x86, arm64) behave similarly enough that we can just test one.

arm64 is the only arch targeted by macOS these days, but it still makes sense to test x86 which wan discountinued in Jan 2023 according to [Wikipedia](https://en.wikipedia.org/wiki/Mac_transition_to_Apple_silicon), so I'm adding an env rather than changing the one we have.

Also note that [jobs relying on the moving tag `macos-latest` are being migrated](https://github.blog/changelog/2024-04-01-macos-14-sonoma-is-generally-available-and-the-latest-macos-runner-image/) to amr64, so I'm setting `macos-12` and `macos-14` explicitly (which is the documented way to request x86 and amr64 runners, respectively).

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
We're getting two failures, which are exactly #16319 and #16320 so this job already proved useful.

**Comment by pllim** [comment]
We were not even able to test OSX ARM64 wheels until recently, but now that we can, I guess does not hurt. But I would want @astrofrog and/or @saimn's inputs on whether a new CI job is really necessary here. As it stands, I think we would have caught this in the nightly wheel build (the scheduled ones, e.g., https://github.com/astropy/astropy/actions/runs/8778293486) but given transient nature, maybe that is not often enough, but then again, is scipy BLAS issues really our problem?

File: .github/workflows/ci_workflows.yml [hunk 1]
```diff
@@ -85,9 +85,15 @@ jobs:
...
 
-        - name: Python 3.11 with all optional dependencies (MacOS X)
+        - name: Python 3.11 with all optional dependencies (MacOS X, arm64)
           macos: py311-test-alldeps
           posargs: --durations=50 --run-slow
+          runs-on: macos-14
+
+        - name: Python 3.11 with all optional dependencies (MacOS X, x86)
+          macos: py311-test-alldeps
+          posargs: --durations=50 --run-slow
```
Comment (saimn) [inline .github/workflows/ci_workflows.yml:95]: We probably don't need `--durations=50 --run-slow` on both ?

**Comment by saimn** [comment]
Moving to the arm64 job is a good thing I think since that's what osx users are more and more using, and it's a different architecture. Not sure if we need to keep the x86 one here though, maybe it could be moved to cron jobs ?

**Comment by neutrinoceros** [comment]
> is scipy BLAS issues really our problem ?

Yes and no:
- I'm not yet sure were the problem is originated (BLAS is only my primary suspect)
- this particular instance is probably symptomatic of an edge case that is exercised in astropy's test suite and not in scipy's. Ultimately the (reduced) test should be upstream, but that's one of the reasons we need to test how our dependencies integrate: we can't expect them to already cover 100% of our own use case in their test suite (maybe in an ideal world though).
- not having this (non exotic) arch part of CI means that we're effectively relying on devs and users to discovers these issues, and I would really like CI to catch such defects before users (or I) do.

> Not sure if we need to keep the x86 one here though, maybe it could be moved to cron jobs ?

sounds reasonable to me.

File: .github/workflows/ci_workflows.yml (hunk 1 above)
Comment (neutrinoceros) [reply .github/workflows/ci_workflows.yml:95]: Good point. I'll update it when we know wether your idea of moving x86 to cron gets traction. Thanks !

**Comment by neutrinoceros** [comment]
good point, I didn't think of that. Thanks !

**Comment by neutrinoceros** [comment]
Looks like we just went through the migration process, so `macos-latest` is now arm64 by default. Since we still have a couple failures, I'll open another PR to move back to x86 to stabilize main, and I'll rebase this one.

**Comment by pllim** [comment]
I merged #16330 which would have to be reverted here along with fixes for if still necessary:

* https://github.com/astropy/astropy/issues/16319
* https://github.com/astropy/astropy/issues/16320

I think we should wait for scipy to sort it out first. Also see #16329 for an unmerged attempt.

**Comment by neutrinoceros** [comment]
For now I'm working on producing a useful reproducer for scipy folks (the full-fledge astropy test involves several thousands lines of code so it's basically impossible to debug). I'm getting there !

**Comment by neutrinoceros** [comment]
I rebased to fix conflicts and resolve existing suggestions. CI is still expected to fail.

**Comment by neutrinoceros** [comment]
Unless the fix in scipy is backported, this will wait for scipy 1.14.0. See https://discuss.scientific-python.org/t/proposed-release-schedule-for-scipy-1-14-0/ for the proposed release schedule

**Comment by pllim** [comment]
Superseded by #16487
